### ISSUE TITLE
Moving cursor while composing should stay in composing region

### ISSUE THREAD

**Issue by gnprice** [flutter/flutter#122490]
When there's an active composing region for a multi-step IME (like for Chinese or Japanese), if the user moves the cursor (aka the caret) around, it should stay clamped within the composing region. This is what iOS and Android do for native text inputs. (With a few exceptions, that is; details below.)

This applies to moving the cursor by tapping inside the text field, and on iOS applies also to using the "floating cursor" feature by long-pressing the space bar and dragging around.

## Steps to Reproduce

1. Go to a text input field in a Flutter app, such as the code sample below, on iOS or Android.
2. Select (for example) a Japanese kana keyboard. Type some text, for example "あかさあかさあかさ". Hit the big "確定" key on iOS, or "enter" key with Gboard on Android, to end composing and confirm your input.
3. Put the cursor in the middle of the text you've typed. Type some more text, for example "たなたな". Leave the input in a still-composing state.
4. Try to move the cursor beyond the composing area:
   * Tap in the left part of the text input, before the "たなたな" composing region.
   * Tap in the right part of the text input, after the composing region.
   * On iOS, long-press the big "確定" key until all the keyboard's keys go blank, then drag far to the left or the right.
5. Observe where the cursor lands.
6. Go to a text input field in a native iOS or Android app, and repeat steps 2-5.

**Expected results:**

Steps 2-5 should produce similar results in Flutter as in a native iOS or Android app.

Specifically: moving far to the left should take the cursor only to the start of the composing region, and moving far to the right should take the cursor only to the end of the composing region.

This behavior is helpful if, for example, you want to get back to the end of the composing region and resume typing there after correcting something in the beginning or middle of the region: it means you can just hastily move the cursor toward the end, without needing to be precise, and it'll go exactly to the end.

**Actual results:**

Moving to a point outside the composing region takes the cursor to that point, and ends composing.

This can be pretty annoying if you wanted to go to the start or end, and you miss by even a little bit: if you weren't done composing, you'll likely have to delete and retype the whole composing region again.

(If you move the cursor somewhere within the composing region, instead of outside it, you'll see a different issue:
* #108016
)

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(DemoApp());
}

class DemoApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(title: "Demo app", home: HomePage());
  }
}

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text("Demo app")),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          TextField(maxLines: null),
          ElevatedButton(
              onPressed: () => FocusScope.of(context).unfocus(),
              child: Text('Unfocus'),
          ),
        ],
      ),
    );
  }
}
```

</details>

**Comment by gnprice** [comment]
### iOS and Android native behavior

The expected behavior above — keep the cursor inside the composing region — describes how both iOS and Android behave, modulo a couple of exceptions. In more detail:

* On iOS, using the long-press-on-space "floating cursor" feature, it always behaves this way — I've never succeeded in getting that feature to place the cursor outside the composing region. While you're dragging, the cursor will move fluidly in tandem with your finger (just like it will move outside the bounds of the whole text input); but when you release, it snaps to the nearest point within the composing region.

* On iOS, if you tap inside the text field before the start of the composing region, it always behaves this way, taking the cursor only to the start of the composing region.

  If you try repeatedly to repro this, it's easy to make a double-tap instead, which does something different. But if you wait a second between taps, you can tap before the start of the region many times in a row and the cursor stays inside the region.

* On iOS, if you tap after the end of the composing region, it usually behaves this way. Sometimes it instead places the cursor where you tapped and cancels composing, just like Flutter's current behavior.

  I haven't worked out a definite pattern of when it does so. Conceivably it's a bug, but probably it's some kind of design choice. Potentially we should imitate this wrinkle, once we figure out what the pattern is.

* On Android, with Gboard in Japanese (using the "12-key" kana keyboard), it always behaves this way. But it looks at first as if it doesn't:

  * When you tap outside the composing region, the cursor appears to go where you tapped: it starts being painted there. The composing region stays highlighted, though.

  * However, if you tap another kana key, the kana you typed doesn't appear where the cursor seemed to be. Instead it appears at the beginning or end of the composing region, according to whether you tried to put the cursor before or after the region.

    So the behavior shows that the cursor's real location is at the beginning or end of the composing region. There's a discrepancy in where the cursor is painted, but that seems like a bug in Android and/or Gboard.

  * Similarly if you hit delete, or use the arrow keys, the behavior shows that the cursor's real location was within the composing region, and it was just being painted in the wrong place.

* On Android, with Gboard in Chinese (using a pinyin keyboard), the UI is quite different — the pinyin letters don't appear in the text input at all, but instead just above the keyboard, and the text input only ever shows the completed input after composing. So this question doesn't come up.


In short, both iOS and Android always clamp the cursor to within the composing region when applicable, with two exceptions:
* Android/Gboard has a bug that affects where the cursor appears to be, but not its actual behavior.
* iOS has one case where it behaves differently.
